crates/prune/src/pruner.rs

```diff
@@ -253,13 +263,163 @@ impl<DB: Database> Pruner<DB> {
                     "Pruned receipts"
                 );
             },
+            |_| false,
         )?;
 
         provider.save_prune_checkpoint(
             PrunePart::Receipts,
             PruneCheckpoint { block_number: to_block, prune_mode },
         )?;
 
+        // `PrunePart::Receipts` overrides `PrunePart::ContractLogs`, so we can preemptively
+        // limit their pruning start point.
+        provider.save_prune_checkpoint(
+            PrunePart::ContractLogs,
+            PruneCheckpoint { block_number: to_block, prune_mode },
+        )?;
+
+        Ok(())
+    }
+
+    /// Prune receipts up to the provided block by filtering logs. Works as in inclusion list, and
+    /// removes every receipt not belonging to it.
+    #[instrument(level = "trace", skip(self, provider), target = "pruner")]
+    fn prune_receipts_by_logs(
+        &self,
+        provider: &DatabaseProviderRW<'_, DB>,
+        tip_block_number: BlockNumber,
+    ) -> PrunerResult {
+        // Contract log filtering removes every receipt possible except the ones in the list. So,
+        // for the other receipts it's as if they had a `PruneMode::Distance()` of 128.
+        let to_block = PruneMode::Distance(MINIMUM_PRUNING_DISTANCE)
+            .prune_target_block(
+                tip_block_number,
+                MINIMUM_PRUNING_DISTANCE,
+                PrunePart::ContractLogs,
+            )?
+            .map(|(bn, _)| bn)
+            .unwrap_or_default();
+
+        // Figure out what receipts have already been pruned, so we can have an accurate
+        // `address_filter`
+        let pruned = provider
+            .get_prune_checkpoint(PrunePart::ContractLogs)?
+            .map(|checkpoint| checkpoint.block_number);
+
+        let address_filter =
+            self.modes.contract_logs_filter.group_by_block(tip_block_number, pruned)?;
+
+        // Splits all transactions in different block ranges. Each block range will have its own
+        // filter address list and will check it while going through the table
+        //
+        // Example:
+        // For an `address_filter` such as:
+        // { block9: [a1, a2], block20: [a3, a4, a5] }
+        //
+        // The following structures will be created in the exact order as showed:
+        // `block_ranges`: [
+        //    (block0, block8, 0 addresses),
+        //    (block9, block19, 2 addresses),
+        //    (block20, to_block, 5 addresses)
+        //  ]
+        // `filtered_addresses`: [a1, a2, a3, a4, a5]
+        //
+        // The first range will delete all receipts between block0 - block8
+        // The second range will delete all receipts between block9 - 19, except the ones with
+        //     emitter logs from these addresses: [a1, a2].
+        // The third range will delete all receipts between block20 - to_block, except the ones with
+        //     emitter logs from these addresses: [a1, a2, a3, a4, a5]
+        let mut block_ranges = vec![];
```

Comment:
nit: let's store ranges in this vec instead of two bound elements? So it'll be a `Vec<(RangeInclusive<BlockNumber>, usize)>`